### DISCUSSION TITLE
usage of regex - sometimes I don't get the desired result, how to bring light on this black box?

### DISCUSSION THREAD

**Discussion by aisbergde** [schemacrawler/SchemaCrawler#439] [General]
SchemaCrawler is based on regex, but to get the right result is often a question of trial and error.

One finding is, that in some cases I have to use " (maybe also ' works) to get the expression into the parameter.

This doesn't work:

```
schemacrawler --url=jdbc:sqlserver://localhost;instancename=sql2019;databaseName=dhw_self; --schemas=dhw\_self\..* --tables=(?i)(dhw\_self\.repo\.generatorusp)|(dhw\_self\.repo\.generatoruspparameter)|(dhw\_self\.repo\.generatoruspstep) --info-level=standard --command=list --user=readonly --password=readonly
```

This works:

```
schemacrawler --url=jdbc:sqlserver://localhost;instancename=sql2019;databaseName=dhw_self; --schemas=dhw\_self\..* --tables="(?i)(dhw\_self\.repo\.generatorusp)|(dhw\_self\.repo\.generatoruspparameter)|(dhw\_self\.repo\.generatoruspstep)" --info-level=standard --command=list --user=readonly --password=readonly
```

### Often the resulting set from the expression is bigger than the expected result

The problem concerns the shell in particular, but not only

some examples:

`grep --grep-columns=dhw\_self\.repo\.RepoObject\..*`
or
`grep --grep-columns="dhw\_self\.repo\.RepoObject\..*"`

In my naive imagination this should limit the object to `dhw_self.repo.RepoObject`, but the result is: I get  **all objects** starting with 'dhw_self.repo.RepoObject', not only the one object. I get half the database because many objects start with "RepoObject".

#427

in the shell, when I use `limit --schemas=WideWorldImporters\..*` I get not only the schema WideWorldImporters, but I get also some object from WideWorldImportersDW. But it works when not in the shell.

schemacrawler.config.properties
When I try to filter out the shema `sys` using this:
`schemacrawler.table.pattern.exclude=.*\.sys\..*`
this will filter not only objects in the schema `sys` (I try to filter objects containing ".sys."), but also object containing `sys` anywhere, for example in the schema `repo_sys`
Idea was to exclude all system objects:
`schemacrawler.table.pattern.exclude=.*\.sys\..*|.*\.INFORMATION_SCHEMA\..*`
but I can't use this filter because it is filtering too much

### black box

In other words, I try to find some regex and I test it in https://regex101.com/ or other places. And there I get the desired result. But in SchemaCrawler sometimes I get the expected result, but sometimes not. This is a very "black box" for me.

- It depends on: am I in the shell or not
- maybe the used scripting engine could also influence the result

Are there ways to shed some more light on this black box?
Could there be ways to see the actual regex expression applied?

**Comment by sualeh** [comment]
@aisbergde - please use SchemaCrawler 16.14.2
